Add Knowledge Base docs (#13)
optakt/optakt.github.io: content/knowledge_base/consensus/avalanche/specification.md

# Avalanche

> An internet-scale electronic payment system, evaluated in a large scale deployment, constructed with Snow family of algorithms. Avalance's tested confirmation latency goes down to 1.35 sec. System's bottleneck is transaction verification.

Reviewer: ```suggestion
An internet-scale electronic payment system, evaluated in a large scale deployment, based on the Snow family of algorithms. Avalanche's tested confirmation latency goes down to 1.35 sec. The system's bottleneck is transaction verification.
```

I'm curious, where is this 1.35s metric coming from?

Author: Seems like I need to elaborate on this,

Author: The text changes are added https://github.com/optakt/optakt.github.io/pull/13/commits/0a1cb902f3417c3bc20f9197df71d3d65a1b6421#diff-69f45991c0732c6f5ff297a5a86ffa91662d50369fc8057b3875361f5b80a2fcR3-R7
I removed the 1.35ms reference because it is a result from a test of the Avalanche network that has a quite long description for the initial specifications and its place is not here. Its from the white paper, I plan to add it in a conclusion section lately

TODO: elaborate more on the introduction

## Characteristics

Avalanche network operates with multiple single-decree instances (nodes), meaning that each instance can make a decision on only one value. Nodes use Snowball algorithm for the process of making a decision. (TODO: add a reference) In the case of a payment system and Avalanche, the Snowball logic used to determine a value is a process of deciding on a transaction being valid or not. 

The protocol maintains the set of all known transactions using a dynamic, append-only Directed Acyclic Graph (DAG) structure. (TODO: add reference)
There are two main positives from this:
 * DAG helps streamlining the path on which there are no conflicting transactions - a single vote on a DAG vertex implicitly votes for all transactions on the path to the genesis vertex
 * using the same principle as BTC, the DAG tree renders past decisions on the path of transactions as more and more difficult to undo without approval of the correct nodes
Genesis block concept is implemented with the DAG having a single sink - "genesis vertex".

> When a new transaction is created its parents are defined in the DAG structure as its ancestors, forming the edges with their newly created decendant. The parent-child relationship in DAG is not mandatory for a child transaction may not have any relation to its parents transactions outputs. It can actually spend funds received in another transactions from its "ancestor set". The "ancestor set" is the set of all transactions reachable via parent edges back in history (TODO: add reference to the quote). Oppositely - the term "progeny" means all existing (and potentially existing) children transactions and their children.

Reviewer: ```suggestion
When a new transaction is created, its parents are defined in the DAG structure as its ancestors, forming the edges with their newly created descendant.
The parent-child relationship in DAG is not mandatory for a child transaction may not have any relation to its parents transactions outputs.
It can actually spend funds received in other transactions from its "ancestor set".
The "ancestor set" is the set of all transactions reachable via parent edges throughout history (TODO: add reference to the quote).
Oppositely — the term "progeny" means all existing (and potentially existing) children transactions and their children.
```

This is true for the whole document, but we basically try to keep one sentence per line, so that git diffs are more digestible when we make changes to markdown documents in the future, and this is also helpful for PR reviews.

Author: Right, I'll go over all the text created up to now and update it; thanks!

Author: Put the sentences at separate lines here and in the rest of the texts: https://github.com/optakt/optakt.github.io/pull/13/commits/0a1cb902f3417c3bc20f9197df71d3d65a1b6421#diff-69f45991c0732c6f5ff297a5a86ffa91662d50369fc8057b3875361f5b80a2fcR24-R28
Thanks!

> Main purpose of a consensus protocol is to avoid inclusion of conflicting transactions into the ledger. Even though a conflicting transaction characteristics are defined on an application level, the notion of conflict can be abstracted in order to define the conflicting set. In Avalanche every transaction belongs to a conflict set, meaning every transaction is a part of a set of transactions that contains one (if the set is a singleton set - the conflicting transactions will be potentially created or the singleton set contains a single virtuos transaction) or more transactions that are invalid towards each other existence.

Reviewer: ```suggestion
The main purpose of a consensus protocol is to avoid the inclusion of conflicting transactions into the ledger.
Even though a conflicting transaction's characteristics are defined on an application level, the notion of conflict can be abstracted in order to define the conflicting set.
In Avalanche, every transaction belongs to a conflict set, meaning every transaction is a part of a set of transactions that contains one (if the set is a singleton set, the conflicting transactions will be potentially created or the singleton set contains a single virtual transaction) or more transactions that are invalid towards each other existence.
```

I guess you meant virtual transaction?

Author: No, it was just unclear, updated this part, IMO is clear now, WDYT?
https://github.com/optakt/optakt.github.io/pull/13/commits/0a1cb902f3417c3bc20f9197df71d3d65a1b6421#diff-69f45991c0732c6f5ff297a5a86ffa91662d50369fc8057b3875361f5b80a2fcR32-R34

Reviewer: It's much better yep :) Thanks!

> Following the purpose of reaching a consensus - only one transaction from any conflict set can be added to the DAG tree. 

Reviewer: ```suggestion
Following the purpose of reaching a consensus — only one transaction from any conflict set can be added to the DAG tree. 
```

I'm not sure what you mean by this sentence 🤔

Author: Updated it to be more understandable https://github.com/optakt/optakt.github.io/pull/13/commits/0a1cb902f3417c3bc20f9197df71d3d65a1b6421#diff-69f45991c0732c6f5ff297a5a86ffa91662d50369fc8057b3875361f5b80a2fcR34-R35
Thanks!

> Avalanche instantiates a Snowball instance for each conflict set. Avalanche treats the concept of repeated queries and multiple counters from Snowball taking advantage of DAG structure.Specifically, when a transaction T is queried, all of its ancestors reachable through the DAG edges from it are implicitly part of the query. Consequently, when a node responds positively to query of T only if T and all of its ancestors are the preferred option in their respective conflict sets.
> If more than a threshold of responders vote positively, the transactions gets a 'chit'. (TODO: add reference)

Reviewer: ```suggestion
Avalanche instantiates a Snowball instance for each conflict set.
Avalanche treats the concept of repeated queries and multiple counters from Snowball taking advantage of DAG structure.
Specifically, when a transaction `T` is queried, all of its ancestors reachable through the DAG edges from it are implicitly part of the query.
Consequently, nodes would respond positively to a query for transaction `T` only if `T` and all of its ancestors are the preferred option in their respective conflict sets.
If more than a given threshold of responders vote positively, the transactions gets a _chit_. (TODO: add reference)
```

Author: Updated the sentences + added references https://github.com/optakt/optakt.github.io/pull/13/commits/0a1cb902f3417c3bc20f9197df71d3d65a1b6421#diff-69f45991c0732c6f5ff297a5a86ffa91662d50369fc8057b3875361f5b80a2fcR36-R42
Thanks!

# Avalanche

An internet-scale electronic payment system, evaluated in a large scale deployment, constructed with Snow family of algorithms. Avalance's tested confirmation latency goes down to 1.35 sec. System's bottleneck is transaction verification.
TODO: elaborate more on the introduction

## Characteristics

> Avalanche network operates with multiple single-decree instances (nodes), meaning that each instance can make a decision on only one value. Nodes use Snowball algorithm for the process of making a decision. (TODO: add a reference) In the case of a payment system and Avalanche, the Snowball logic used to determine a value is a process of deciding on a transaction being valid or not. 

Reviewer: ```suggestion
The Avalanche network operates with multiple single-decree instances (nodes), meaning that each instance can make a decision on only one value. Nodes use the Snowball algorithm for their decision-making process. In the case of a payment system and Avalanche, the Snowball logic used to determine a value is the process of deciding whether a transaction is valid. 
```

IMO the references within sentences should be included in this PR, as it also makes it easier to review properly and we might forget them in the future otherwise since they're hidden within paragraphs. WDYT?

Author: Yes, I'll go over them and add them, thanks!

Author: This text has been updated, added references as well
https://github.com/optakt/optakt.github.io/pull/13/commits/0a1cb902f3417c3bc20f9197df71d3d65a1b6421#diff-69f45991c0732c6f5ff297a5a86ffa91662d50369fc8057b3875361f5b80a2fcR10-R16
Thanks!

The protocol maintains the set of all known transactions using a dynamic, append-only Directed Acyclic Graph (DAG) structure. (TODO: add reference)
There are two main positives from this:
 * DAG helps streamlining the path on which there are no conflicting transactions - a single vote on a DAG vertex implicitly votes for all transactions on the path to the genesis vertex
 * using the same principle as BTC, the DAG tree renders past decisions on the path of transactions as more and more difficult to undo without approval of the correct nodes

> Genesis block concept is implemented with the DAG having a single sink - "genesis vertex".

Reviewer: ```suggestion
The protocol maintains the set of all known transactions using a dynamic, append-only Directed Acyclic Graph (DAG) structure.

There are two main positives from this:

 * DAG helps streamlining the path on which there are no conflicting transactions — a single vote on a DAG vertex implicitly votes for all transactions on the path to the genesis vertex
 * using the same principle as BTC, the DAG tree renders past decisions on the path of transactions as more and more difficult to undo without approval of the correct nodes
 
Genesis block concept is implemented with the DAG having a single sink - "genesis vertex".
```

Also I'm not sure how but I feel like we should rephrase the last sentence, WDYT?

Author: Updated the description, IMO is more clear now https://github.com/optakt/optakt.github.io/pull/13/commits/0a1cb902f3417c3bc20f9197df71d3d65a1b6421#diff-69f45991c0732c6f5ff297a5a86ffa91662d50369fc8057b3875361f5b80a2fcR21-R22
WDYT?
